### ISSUE TITLE
Webserver security settings do net get recognized

### ISSUE THREAD

**Issue by mmattel** [owncloud/core#15342]
Ubuntu 14.04, nginx 1.7.11, MySQL, Browser W7/Opera 28

@LukasReschke 

Installing via "sudo ocdev base --dir /xxxx" a new instance of oC from master for testing and finalizing the installation.

I have configured in my sites server ssl block:
        add_header Strict-Transport-Security "max-age=63072000; includeSubDomains; always; preload";
        add_header X-Content-Type-Options nosniff;
        add_header X-Frame-Options "SAMEORIGIN";
 
But after installing, I get warnings at the admin page about missing of these settings

![image](https://cloud.githubusercontent.com/assets/3321281/6925267/9b344be8-d7de-11e4-9602-b9853a9ce4f6.png)

This belongs to: core/js/tests/specs/setupchecksSpec.js

If I do:  sudo curl -I https://myserver -k

HTTP/1.1 200 OK
Server: nginx
Date: Tue, 31 Mar 2015 17:22:43 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
X-Powered-By: PHP/5.5.9-1ubuntu4.7
Set-Cookie: xxxx; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src *; font-src 'self' data:; media-src *; connect-src *
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: Sameorigin
X-Robots-Tag: none
Strict-Transport-Security: max-age=63072000; includeSubDomains; always; preload
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN

(interestingly I get nosniff and sameorigin twice but I never configured this twice, was looking for it with grep but anyway...)

**Comment by LukasReschke** [comment]
Probably the check doesn't like receiving the header twice. Let me take a look at that later in this dev cycle.

**Comment by mmattel** [comment]
I was searching in my nginx directory recursively for

1. "SAMEORIGIN" and
2. "Sameorigin"

and I only got results for "SAMEORIGIN", exactly what I set in the sites config

**Comment by LukasReschke** [comment]
Yes. It is set via PHP as well unless https://github.com/owncloud/core/blob/691c353eeda077be0fcca1e3839265815a650a90/.htaccess#L16 is set - I'm not sure how you can set this in NGinx but you might try to do this :-)

**Comment by mmattel** [comment]
In nginx, you set this once in the ssl definition of the site.

from oC .htaccess as from the link above

    Header set X-Content-Type-Options "nosniff"
    Header set X-XSS-Protection "1; mode=block"
    Header set X-Robots-Tag "none"
    Header set X-Frame-Options "SAMEORIGIN"

I do not set the X-XSS and the X-Robots-Tag header in nginx, but it is part of the result of the curl response... But maybe this is set i nginx by default and does not need to be defined extra - will take a look at.

The interresting thing is the following:
From what I have read, you can not create a global ssl setting containing add_header directives in nginx and then just add a site specific add_header directive. The first site specific add_header directive kills all global ones. Therefore you need to define for all sites you have the add_header directives in the ssl part of the particular site definition. Knowing that, I wonder how this double response can take place.

What I can do to proof is, to just rename in .htaccess "SAMEORIGIN" to"sAMEORIGIN" and see if this gets propagated.

Can you please tell me what you exactly mean with: is set via PHP unless...?

**Comment by mmattel** [comment]
From: core/lib/private/response.php#L233

		// Send fallback headers for installations that don't have the possibility to send
		// custom headers on the webserver side
		if(getenv('modHeadersAvailable') !== 'true') {
			header('X-XSS-Protection: 1; mode=block'); // Enforce browser based XSS filters
			header('X-Content-Type-Options: nosniff'); // Disable sniffing the content type for IE
			header('X-Frame-Options: Sameorigin'); // Disallow iFraming from other domains
			header('X-Robots-Tag: none'); // https://developers.google.com/webmasters/control-crawl-index/docs/robots_meta_tag

``getenv('modHeadersAvailable')`` can only return true when using Apache but not when eg using nginx or others. A solution would be a config.php entry (eg ``security_headers_set``) where the admin declares manually that he defined these headers in his webserver properly - then no double sending will be done.

It is principially a good idea to make the code double response proof, you never know, but double sending should be avoided.

**Comment by mmattel** [comment]
Small note to:

core/lib/private/response.php#L233 and https://github.com/owncloud/core/blob/691c353eeda077be0fcca1e3839265815a650a90/.htaccess#L16

Maybe a remark in the code blocks pointing to each other will remind people to not forget the other side...

**Comment by mmattel** [comment]
Good news, at least some parts of the issue is on the right way to get solved :smiley:

I maybe found a possible solution to use a equivalent of the apache ``setenv`` directive for nginx. For other webservers I do not know.

It is important that I have enabled fast_cgi in my nginx config, which is btw suggested in the admin/installation manual (=possible doc update....)

http://stackoverflow.com/questions/8098927/nginx-variables-similar-to-setenv-in-apache
http://stackoverflow.com/questions/8551592/setting-env-fka-http-env-vars-with-nginx-php-fpm

``fastcgi_param modHeadersAvailable true;``

Whyever, curl now responds imho properly, also in exact the same sort order I have put the add_header directives in, but oC still reports that the security parametes are not set.

sudo curl -I https://myserver -k
HTTP/1.1 200 OK
Server: nginx
Date: Wed, 01 Apr 2015 07:22:25 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
X-Powered-By: PHP/5.5.9-1ubuntu4.7
Set-Cookie: xxxx; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src *; font-src 'self' data:; media-src *; connect-src *
Strict-Transport-Security: max-age=63072000; includeSubDomains; always; preload;
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN

**Comment by mmattel** [comment]
Idea (after fixing the issue...):

if I look at the curl response, I see that there is an output for the server type ``Server: nginx``
As stuff is handled fo Apache automatically via .htaccess but not for other webservers, an extended warning based on the webserver used would maybe help.

Would it be possible to add at the check the webserver type and based on the type a proper hint about how to manage this properly pointing to the documentation with an hint what to look for?

**Comment by DeepDiver1975** [comment]
@LukasReschke I'm moving this to 8.2

**Comment by RandolfCarter** [comment]
Just encountered the "Strict-Transport-Security" check (because it is included in oc 8.1b2). From my experiments it seems that the check is expecting a wrong format: It expects the Header to look like this (with such a header in the apache config, the warning is not shown anymore):
```
Header always set Strict-Transport-Security "31536000; includeSubDomains"
```
when in fact it should look like this:
```
Header always set Strict-Transport-Security "max-age=31536000; includeSubDomains"
```
(notice the added max-age).
A quick fix could be:
```
--- core/js/setupcheck.js.old      2015-05-22 00:51:22.784632554 +0200
+++ core/js/setupchecks.js.new      2015-05-22 00:49:23.728170921 +0200
@@ -153,6 +153,7 @@
                                                } else {
                                                        transportSecurityValidity = transportSecurityValidity.substring(8);
                                                }
+                                        transportSecurityValidity = transportSecurityValidity.split('=').pop();
                                        }

                                        if(isNaN(transportSecurityValidity) || transportSecurityValidity <= 2678399) {
```
But I would leave the final implementation of that to more capable hands, my JS is a bit rusty ;)
